sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslUdfUdafTest.java

```diff
@@ -133,6 +137,40 @@ public void testUdf() throws Exception {
     pipeline.run().waitUntilFinish();
   }
 
+  /** test auto-registered UDF/UDAF. */
+  @Test
+  public void testAutoUdfUdaf() throws Exception {
+    Schema resultType =
+        Schema.builder().addInt32Field("f_int2").addInt32Field("squarecubicsum").build();
+
+    Row row = Row.withSchema(resultType).addValues(0, 4890).build();
+
+    String sql =
+        "SELECT f_int2, squaresum(cubic(f_int)) AS `squarecubicsum`"
+            + " FROM PCOLLECTION GROUP BY f_int2";
+    PCollection<Row> result =
+        boundedInput1.apply("testUdaf", SqlTransform.query(sql).autoRegisterUdfUdaf());
+
+    PAssert.that(result).containsInAnyOrder(row);
+    pipeline.run().waitUntilFinish();
+  }
+
+  /** Auto register for test. */
+  @AutoService(UdfUdafRegister.class)
+  public static class UdfUdafRegisterTest implements UdfUdafRegister {
```

Comment:
why `UdfUdafRegisterTest` is a static class without a static variable/function? Why to make it static?

Reply:
Static nested classes have nothing to do with static members, it's about access to the outer class' non-static variables (obviously): https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html

Reply:
Static in this context means that it doesn't reference the the parent class. If it wasn't static it would only be accessible on an instantiation of `BeamSqlDslUdfUdafTest`, which means the service loader couldn't find it.

Reply:
Thanks!